geode-apis-compatible-with-redis/src/integrationTest/java/org/apache/geode/redis/internal/executor/sortedset/AbstractZRemIntegrationTest.java

```diff
@@ -0,0 +1,178 @@
+
+  private static final int REDIS_CLIENT_TIMEOUT =
+      Math.toIntExact(GeodeAwaitility.getTimeout().toMillis());
+  private static final String SORTED_SET_KEY = "ss_key";
+  private static final int INITIAL_MEMBER_COUNT = 5;
+
+  @Before
+  public void setUp() {
+    jedis = new JedisCluster(new HostAndPort("localhost", getPort()), REDIS_CLIENT_TIMEOUT);
+  }
+
+  @After
+  public void tearDown() {
+    flushAll();
+    jedis.close();
+  }
+
+  @Test
+  public void zRemThrowsIfTooFewArguments() {
+    assertAtLeastNArgs(jedis, Protocol.Command.ZREM, 2);
+  }
+
+  @Test
+  public void zRemThrowsIfGivenOnlyKey() {
+    assertThatThrownBy(() -> jedis.zrem("key")).hasMessageContaining("wrong number of arguments");
+  }
+
+  @Test
+  public void zRemThrowsErrorIfKeyIsNotASortedSet() {
+    String key = "key";
+    String member = "member1";
+    jedis.sadd(key, member);
+
+    assertThatThrownBy(() -> jedis.zrem(key, member))
+        .hasMessageContaining(ERROR_WRONG_TYPE);
+  }
+
+  @Test
+  public void zRemDoesNotRemoveNonExistingMember() {
+    Map<String, Double> map = makeMemberScoreMap(INITIAL_MEMBER_COUNT);
+    addToSortedSet(map);
+
+    String nonExistingMember = "nonExisting";
+    long result = jedis.zrem(SORTED_SET_KEY, nonExistingMember);
+
+    assertThat(result).isEqualTo(0);
+  }
+
+  @Test
+  public void zRemCanRemoveAMemberInASortedSet() {
+    Map<String, Double> map = makeMemberScoreMap(INITIAL_MEMBER_COUNT);
+    Set<String> keys = map.keySet();
+    addToSortedSet(map);
+
+    String memberToRemove = baseName + 1;
+    Long removed = jedis.zrem(SORTED_SET_KEY, memberToRemove);
+    assertThat(removed).isEqualTo(1);
+
+    for (String member : keys) {
+      Double score = jedis.zscore(SORTED_SET_KEY, member);
+      if (member.equals(memberToRemove)) {
+        assertThat(score).isNull();
+      } else {
+        assertThat(score).isNotNull();
+      }
+    }
+    assertThat(jedis.exists(SORTED_SET_KEY)).isTrue();
+  }
+
+  @Test
+  public void zRemRemovesKeyIfRemoveAllMembersInASortedSet() {
+    Map<String, Double> map = makeMemberScoreMap(INITIAL_MEMBER_COUNT);
+    Set<String> keys = map.keySet();
+    addToSortedSet(map);
+
+    String[] membersToRemove = new String[keys.size()];
+    Long removed = jedis.zrem(SORTED_SET_KEY, keys.toArray(membersToRemove));
+    assertThat(removed).isEqualTo(keys.size());
+
+    for (String member : keys) {
+      Double score = jedis.zscore(SORTED_SET_KEY, member);
+      assertThat(score).isNull();
+    }
+    assertThat(jedis.exists(SORTED_SET_KEY)).isFalse();
+  }
+
+  @Test
+  public void zRemCanRemovesMembersConcurrentlyInASortedSet() {
+    int membersCount = 1000;
+    Map<String, Double> map = makeMemberScoreMap(membersCount);
+    addToSortedSet(map);
+
+    AtomicInteger totalRemoved = new AtomicInteger();
+    new ConcurrentLoopingThreads(2,
+        (i) -> doZRem(map, totalRemoved),
+        (i) -> doZRem1(membersCount, totalRemoved)).run();
```

Comment:
The issue I see with the above methods is that each thread loop on the same member in order. The second thread always trying to remove the first thread just removed (understandably it is how we use the synchronization to prevent concurrent modification on the sorted set).

Reply:
I think that's kind of the point though... that we have two threads trying to remove something from the same set at the same time. We want to make sure we don't hit and concurrent modification exceptions or report we've removed the same thing twice. We also can't be sure which thread will run/complete first.

Reply:
The order that the threads execute the commands in `ConcurrentLoopingThreads` is not deterministic. So either one could be 'first' to remove a given entry.

Reply:
At least that is what I see from running the test with above mentioned methods. The second one is always blocked on the synchronization code in Redis -- when that thread is unblocked and working on the key. The first thread is blocked on the key. At least this is what I found. I think we at least want to test that both threads can remove some of the members in the sorted set.

Reply:
Ideally, the function you pass into `ConcurrentLoopingThreads` shouldn't be doing its own loop. As it is currently set up, the CLT will only loop twice and the second time everything is already gone. Perhaps one slight adjustment to @sabbey37's suggestion is to replace `run()` with `runInLockstep()`. This means that for every iteration of the CLT loop, all threads will wait for completion of a given iteration before moving to the next one. The idea is that it increases the possibility of concurrency errors if they exist.

Reply:
Removed the own loop in the methods.